### PR TITLE
Publish anonymous gist from command line

### PR DESCRIPTION
**Pull request by microsoftsam**
Adding feature to publish a project to an anonymous gist.
To test: 
```
mkdir testproject; cd testproject
pxt init
pxt gist
```

Command line output: 
Publishing project to gist anonymously.
Uploading....
Success, your gist url is https://gist.github.com/527cb76153bb38b5c86aac5697619a6e
You can load your published project at https://pxt.microbit.org/#pub:gh/gists/527cb76153bb38b5c86aac5697619a6e

Also fixed issue with package install, see github.ts

### CODE REVIEW

File: cli/cli.ts [hunk 1]
```diff
@@ -3423,6 +3423,52 @@ function checkDocsAsync(...args: string[]): Promise<void> {
     return Promise.resolve();
 }
 
+function publishGistCoreAsync(): Promise<void> {
+    console.log("Publishing project to gist anonymously.");
+    return Promise.resolve()
```
Comment (pelikhan) [inline cli/cli.ts]: Do we need this line?

File: cli/cli.ts [hunk 2]
```diff
@@ -3423,6 +3423,52 @@ function checkDocsAsync(...args: string[]): Promise<void> {
...
+        .then(() => {
+            return mainPkg.loadAsync()
+        })
+        .then(() => {
+            console.log("Uploading....")
+            let files: string[] = mainPkg.getFiles()
+            let pxtConfig = mainPkg.config;
+            let filesMap: Map<{}> = {};
```
Comment (pelikhan) [inline cli/cli.ts]: ... ``Map<{ content:string;}>ˋ`

File: cli/cli.ts [hunk 3]
```diff
@@ -3423,6 +3423,52 @@ function checkDocsAsync(...args: string[]): Promise<void> {
...
+
+            files.forEach((fn) => {
+                let fileContent = fs.readFileSync(fn, "utf8");
+                if (fileContent) {
+                    filesMap[fn] = {
+                        "content": fileContent
+                    }
+                } else {
+                    // Cannot publish empty files, go through and remove empty file references form pxt.json
```
Comment (pelikhan) [inline cli/cli.ts]: form -> from

File: cli/cli.ts [hunk 4]
```diff
@@ -3423,6 +3423,52 @@ function checkDocsAsync(...args: string[]): Promise<void> {
...
+                    if (pxtConfig.files && pxtConfig.files.indexOf(fn) > -1) {
+                        pxtConfig.files.splice(pxtConfig.files.indexOf(fn), 1);
+                    } else if (pxtConfig.testFiles && pxtConfig.testFiles.indexOf(fn) > -1) {
+                        pxtConfig.testFiles.splice(pxtConfig.testFiles.indexOf(fn), 1);
+                    }
+                }
+            })
+            // Add pxt.json
+            filesMap['pxt.json'] = {
+                "content": JSON.stringify(pxtConfig)
```
Comment (pelikhan) [inline cli/cli.ts]: Make it 4 spaces to make it more readable

**Comment by pelikhan** [comment]
Could we support non-anonymous gists and updates?

File: cli/cli.ts (hunk 1 above)
Comment (microsoftsam) [inline cli/cli.ts]: Nope, mainPkg.loadAsync returns our promise.

File: cli/cli.ts (hunk 4 above)
Comment (microsoftsam) [inline cli/cli.ts]: Sure.

**Comment by microsoftsam** [comment]
Yes, but we might want to consider pulling in a dependency on a tool that already does this. 
For example: https://github.com/defunkt/gist (MIT).

Reason for this, is we'd need to deal with github login and two factor authentication.

**Comment by pelikhan** [comment]
We should also support the scenario of a signed in user and allowing to patch an existing gist.

**Comment by pelikhan** [comment]
We don't need to deal with oauth if the use provides a token (which can be generated from the GitHub UI). Seems that this tool is written in ruby so we wouldn't want to take on that dependency.

Get Outlook for iOS<https://aka.ms/o0ukef>

________________________________

**Comment by microsoftsam** [comment]
Okay, so we should
- use the git auth token if it's there and pass it with the request
- if there's not git auth token, ask the user for their username and password (just like any private git command). 

It just seems like we're reinventing the wheel a little here, I'll keep digging. 
In the meantime, looks like this PR is good to go.

**Comment by pelikhan** [comment]
Let's not ask the user/password (with 2FA things are more complicated). Uses gives a token a nothing.

Need to store gist id in PXT.json to be able to update it.

Get Outlook for iOS<https://aka.ms/o0ukef>

________________________________

**Comment by pelikhan** [comment]
PR good to go

**Comment by microsoftsam** [comment]
Great. Thanks!

**Pull request merged**